org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/WorkspaceExecuteCommandHandler.java

```diff
@@ -0,0 +1,141 @@
+import org.eclipse.core.runtime.SafeRunner;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.jdt.ls.core.internal.IParticipant;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.lsp4j.ExecuteCommandParams;
+
+public class WorkspaceExecuteCommandHandler {
+
+	private static final String COMMAND = "command";
+
+	private static final String CLASS = "class";
+
+	private static final String ID = "id";
+
+	private static class ParticipantDescriptor {
+
+		private final IConfigurationElement fConfigurationElement;
+
+		private Set<String> fCommandIds;
+
+		private IParticipant fParticipantInstance;
+
+		public ParticipantDescriptor(IConfigurationElement element) {
+			fConfigurationElement = element;
+
+			IConfigurationElement[] children = fConfigurationElement.getChildren(COMMAND);
+			if (children.length > 0) {
+				fCommandIds = new HashSet<>();
+				for (IConfigurationElement child : children) {
+					fCommandIds.add(child.getAttribute(ID));
+				}
+			} else {
+				fCommandIds = Collections.emptySet();
+			}
+
+			fParticipantInstance = null;
+		}
+
+		public Set<String> getCommands() {
+			return fCommandIds;
+		}
+
+		public IParticipant getParticipant() {
+			if (fParticipantInstance == null) {
+				try {
+					Object extension = fConfigurationElement.createExecutableExtension(CLASS);
+					if (extension instanceof IParticipant) {
+						fParticipantInstance = (IParticipant) extension;
+					} else {
+						String message = "Invalid extension to " + IParticipant.EXTENSION_POINT_ID + ". Must implements IParticipant";
+						JavaLanguageServerPlugin.logError(message);
+						return null;
+					}
+				} catch (CoreException e) {
+					JavaLanguageServerPlugin.logException("Unable to create participant ", e);
+					return null;
+				}
+			}
+			return fParticipantInstance;
+		}
+	}
+
+	private static ParticipantDescriptor[] fgContributedParticipants;
+
+	private static ParticipantDescriptor[] getParticipants() {
+		if (fgContributedParticipants == null) {
+			IConfigurationElement[] elements = Platform.getExtensionRegistry().getConfigurationElementsFor(IParticipant.EXTENSION_POINT_ID);
+			fgContributedParticipants = new ParticipantDescriptor[elements.length];
+			for (int i = 0; i < elements.length; i++) {
+				fgContributedParticipants[i] = new ParticipantDescriptor(elements[i]);
+			}
+		}
+		return fgContributedParticipants;
+	}
+
+	/**
+	 * Execute workspace command and invoke language server participant for
+	 * matching command
+	 *
+	 * @param params
+	 *            parameter from the protocol
+	 * @param monitor
+	 * @return execution result
+	 */
+	public Object executeCommand(ExecuteCommandParams params, IProgressMonitor monitor) {
+		if (params == null || params.getCommand() == null) {
+			JavaLanguageServerPlugin.logError("The workspace execute command has empty params or command");
+			return null;
+		}
+
+		ParticipantDescriptor[] correctionProcessors = getParticipants();

```

Comment:
```java
Collection<ParticipantDescriptor> candidates = correctionProcessors.stream().filter(desc -> desc.matches(params.getCommand())).collect(Collectors.toSet()); //no cancellation here but it's super fast so it's ok.

if(candidates.size() > 1) {
   throw "Found multiple participants ({0}) matching command {1}", candidates, params.getCommand()
}
if (monitor.isCanceled()) {
 return null;
}
//execute candidates.iterator().next()
...
```

Reply:
Not sure whether we need throw exception here. All the other handlers mostly log the error/exception information and return default values. I'll do the same for this behavior.